File: internal/adapterproto/codec.go
package adapterproto

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

const MaxFrameSize uint32 = 16 << 20

var (
	ErrEmptyFrame    = errors.New("adapter protocol: empty frame")
	ErrFrameTooLarge = errors.New("adapter protocol: frame too large")
)

func WriteFrame(w io.Writer, value any) error {
	payload, err := json.Marshal(value)
	if err != nil {
		return fmt.Errorf("adapter protocol: encode frame: %w", err)
	}
	if uint64(len(payload)) > uint64(MaxFrameSize) {
		return fmt.Errorf(
			"%w: %d bytes exceeds %d",
			ErrFrameTooLarge,
			len(payload),
			MaxFrameSize,
		)
	}

	var prefix [4]byte
	binary.BigEndian.PutUint32(prefix[:], uint32(len(payload)))

	if err := writeAll(w, prefix[:]); err != nil {
		return fmt.Errorf("adapter protocol: write length: %w", err)
	}
	if err := writeAll(w, payload); err != nil {
		return fmt.Errorf("adapter protocol: write payload: %w", err)
	}
	return nil
}

func ReadFrame(r io.Reader, value any) error {
	var prefix [4]byte
	if _, err := io.ReadFull(r, prefix[:]); err != nil {
		return fmt.Errorf("adapter protocol: read length: %w", err)
	}

	size := binary.BigEndian.Uint32(prefix[:])
	if size == 0 {
		return ErrEmptyFrame
	}
	if size > MaxFrameSize {
		return fmt.Errorf(
			"%w: %d bytes exceeds %d",
			ErrFrameTooLarge,
			size,
			MaxFrameSize,
		)
	}

	payload := make([]byte, int(size))
	if _, err := io.ReadFull(r, payload); err != nil {
		return fmt.Errorf("adapter protocol: read payload: %w", err)
	}
	if err := json.Unmarshal(payload, value); err != nil {
		return fmt.Errorf("adapter protocol: decode frame: %w", err)
	}
	return nil
}

func writeAll(w io.Writer, data []byte) error {
	for len(data) > 0 {
		n, err := w.Write(data)
		if err != nil {
			return err
		}
		if n == 0 {
			return io.ErrShortWrite
		}
		data = data[n:]
	}
	return nil
}

File: examples/rust-tokenring-adapter/src/codec.rs
// Licensed under the Apache License, Version 2.0.

use serde::de::DeserializeOwned;
use serde::Serialize;
use std::error::Error;
use std::fmt;
use std::io::{self, Read, Write};

pub const MAX_FRAME_SIZE: u32 = 16 << 20;

#[derive(Debug)]
pub enum CodecError {
    Io(io::Error),
    Json(serde_json::Error),
    EmptyFrame,
    FrameTooLarge { size: usize, maximum: usize },
}

impl fmt::Display for CodecError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(error) => write!(formatter, "adapter protocol I/O: {error}"),
            Self::Json(error) => {
                write!(formatter, "adapter protocol JSON: {error}")
            }
            Self::EmptyFrame => formatter.write_str("adapter protocol frame is empty"),
            Self::FrameTooLarge { size, maximum } => write!(
                formatter,
                "adapter protocol frame is {size} bytes; maximum is {maximum}"
            ),
        }
    }
}

impl Error for CodecError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Io(error) => Some(error),
            Self::Json(error) => Some(error),
            Self::EmptyFrame | Self::FrameTooLarge { .. } => None,
        }
    }
}

impl From<io::Error> for CodecError {
    fn from(error: io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<serde_json::Error> for CodecError {
    fn from(error: serde_json::Error) -> Self {
        Self::Json(error)
    }
}

pub fn write_frame<W, T>(writer: &mut W, value: &T) -> Result<(), CodecError>
where
    W: Write,
    T: Serialize,
{
    let payload = serde_json::to_vec(value)?;
    let maximum = MAX_FRAME_SIZE as usize;

    if payload.len() > maximum {
        return Err(CodecError::FrameTooLarge {
            size: payload.len(),
            maximum,
        });
    }

    let size = u32::try_from(payload.len()).map_err(|_| CodecError::FrameTooLarge {
        size: payload.len(),
        maximum,
    })?;

    writer.write_all(&size.to_be_bytes())?;
    writer.write_all(&payload)?;
    Ok(())
}

pub fn read_frame<R, T>(reader: &mut R) -> Result<T, CodecError>
where
    R: Read,
    T: DeserializeOwned,
{
    let mut prefix = [0_u8; 4];
    reader.read_exact(&mut prefix)?;

    let size = u32::from_be_bytes(prefix);
    if size == 0 {
        return Err(CodecError::EmptyFrame);
    }
    if size > MAX_FRAME_SIZE {
        return Err(CodecError::FrameTooLarge {
            size: size as usize,
            maximum: MAX_FRAME_SIZE as usize,
        });
    }

    let mut payload = vec![0_u8; size as usize];
    reader.read_exact(&mut payload)?;

    Ok(serde_json::from_slice(&payload)?)
}
